src/org/mozilla/javascript/Hashtable.java

```diff
@@ -0,0 +1,224 @@
+package org.mozilla.javascript;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * This generic hash table class is used by Set and Map. It uses
+ * a standard HashMap for storing keys and values so that we can handle
+ * lots of hash collisions if necessary, and a list to support the iterator
+ * capability. This second one is important because JavaScript handling of
+ * the iterator is completely different from the way that Java does it.
```

Comment:
Nice documentation including the motivation for this impl. HtmlUnit has since a long time it's own impl of Map and Set. Both are using LinkedHashSet for the elements; can you please explain where the difference is that makes this impl necessary.